File: Domain/Entities/Task.cs
namespace ToDoCalendarWeb.Domain;

public class Task : AbstractTrackableEntity
{
    public override Guid Id { get; set; }
    public required string Name { get; set; }
    public string? Description { get; set; }
    public required Guid PeriodId { get; set; }
    public bool IsDraft { get; set; }
    public bool IsCompleted { get; set; }
    public override bool IsTracked { get; set; } = true;
}

File: todocalendarweb.client/src/models/task.ts

export type TaskCreateDTO = {
    readonly name: string
    readonly description?: string
    readonly periodId: string
}

export type TaskSchema = {
    readonly id?: string
    readonly name: string
    readonly description?: string
    readonly periodId: string
    readonly isCompleted: boolean
}

export class Task {

    readonly id: string
    readonly name: string
    readonly description?: string
    readonly periodId: string
    readonly isCompleted: boolean
    readonly schema: TaskSchema

    constructor(id : string, name: string, periodId: string,
        isCompleted: boolean, description?: string) {
        
        this.id = id
        this.name = name
        this.description = description !== undefined ? description : ''
        this.periodId = periodId
        this.isCompleted = isCompleted

        this.schema = {
            ...this
        }
    }

    static createFromSchema(schema : TaskSchema) : Task {
        if(!schema.id)
            throw new Error('ID cannot be empty')

        return new Task(schema.id, schema.name, schema.periodId,
            schema.isCompleted, schema.description)
    }
}